### PR TITLE
Snapshot Blob Disks for SSA

### PR DESCRIPTION
**Pull request by jerryk55**
Add support to Snapshot Blob (non-Managed) disks for SSA.
Previously only Managed Disks were snapshotted.

This PR is one of three in support of a fix to https://bugzilla.redhat.com/show_bug.cgi?id=1463780.
The other two are https://github.com/ManageIQ/manageiq/pull/15960 and https://github.com/ManageIQ/manageiq-smartstate/pull/26

These three PRs should be able to be merged in any order.
@roliveri please review for SSA functionality.  @bronaghs or @djberg96 please review and merge when appropriate.  Thank you.

### CODE REVIEW

File: app/models/manageiq/providers/azure/cloud_manager/vm_or_template_shared.rb [hunk 1]
```diff
@@ -1,9 +1,53 @@
...
+    @vm_object ||= provider_service.get(name, resource_group.name)
+  end
+
+  def os_disk
+    @os_disk ||= vm_object.properties.storage_profile.os_disk
+  end
+
+  delegate :managed_disk?, to: :vm_object
+
+  def snap_name
```
Comment (roliveri) [inline app/models/manageiq/providers/azure/cloud_manager/vm_or_template_shared.rb]: Should the name of this method be more specific, like `ssa_snap_name`?

File: app/models/manageiq/providers/azure/cloud_manager.rb [hunk 2]
```diff
@@ -119,26 +117,32 @@ def vm_reboot_guest(vm, _options = {})
...
+    @connection = connect(options)
+    vm.provider_service(@connection)
+    if vm.managed_disk?
+      vm_create_evm_managed_snapshot(vm)
+    else
+      vm_create_evm_blob_snapshot(vm)
+    end
+  end
+
+  def vm_create_evm_managed_snapshot(vm)
```
Comment (roliveri) [inline app/models/manageiq/providers/azure/cloud_manager.rb]: Someday, we should move all of these `vm_*` methods into the their corresponding VM objects (via mixin or whatever is appropriate). This is a general comment that applies to all providers, and shouldn't be addressed in this PR. When this is addressed, we should probably also change the `evm` in names to something like `ssa`.

File: app/models/manageiq/providers/azure/cloud_manager/vm_or_template_shared.rb (hunk 1 above)
Comment (jerryk55) [reply app/models/manageiq/providers/azure/cloud_manager/vm_or_template_shared.rb]: Definitely.  Changing....

File: app/models/manageiq/providers/azure/cloud_manager.rb (hunk 2 above)
Comment (jerryk55) [reply app/models/manageiq/providers/azure/cloud_manager.rb]: Agreed.  There were suggestions made by @chessbyte on a previous PR in app/models/vm_scan.rb that should probably tackled at the same time.  We should add (an) issue(s) to track this.

File: app/models/manageiq/providers/azure/cloud_manager/vm_or_template_shared/scanning.rb [hunk 3]
```diff
@@ -58,16 +58,6 @@ def requires_storage_for_scan?
...
-        _log.debug("VM #{name} has a Managed Disk - Snapshot required for Scan")
-        return true
-      end
-    rescue => err
-      _log.error("Error Class=#{err.class.name}, Message=#{err.message}")
-    end
-    _log.debug("VM #{name} does not have a Managed Disk - no Snapshot required for Scan")
-    false
+    true
   end
```
Comment (bronaghs) [inline app/models/manageiq/providers/azure/cloud_manager/vm_or_template_shared/scanning.rb]: @jerryk55 - what is the point of this method if it just returns 'true'

**Comment by jerryk55** [comment]
Bronagh,

Different providers may or may not require
Snapshots for SSA.  The default is false.  Prior to the Managed disk change in the hotfix sent out a few days back, Azure required no snapshots.  With this BZ and its associated PRs Azure always requires a snapshot.  

Jerry

Support My Bike Ride To Cure Type 1 Diabetes

http://www2.jdrf.org/site/TR?fr_id=6889&pg=personal&px=1266643

**Comment by jerryk55** [comment]
@miq-bot add_label fine/yes

**Comment by djberg96** [comment]
@bronaghs @jerryk55 is this a candidate for the SupportsFeatureMixin? I see `:snapshots` in there now, along with a few other snapshot related tags.

**Comment by simaishi** [comment]
Fine backport details:

```
$ git log -1
commit 8d71d0a2895df66ae6d02c0a99eb3c1fb5493a98
Author: Bronagh Sorota <bsorota@redhat.com>
Date:   Tue Sep 12 14:02:44 2017 -0400

    Merge pull request #122 from jerryk55/nonmanaged_disk_snapshot_support
    
    Snapshot Blob Disks for SSA
    (cherry picked from commit 29116cc956ae80790e14be488059dc0200770ec6)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1491310
```

**Pull request merged**